### PR TITLE
Updated variants in a few packages

### PR DESCRIPTION
**Pull request by alalazo**
variants: fixed packages reported by @adamjstewart in #4098

### CODE REVIEW

File: var/spack/repos/builtin/packages/matlab/package.py [hunk 1]
```diff
@@ -45,8 +45,19 @@ class Matlab(Package):
...
+        'mode',
+        default='interactive',
+        values=('interactive', 'silent', 'automated'),
+        description='Installation mode (interactive, silent, or automated)'
+    )
+
+    variant(
+        'key',
+        default='',
+        values=lambda x: True,  # Anything goes as a key
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/matlab/package.py]: Why is this line needed?  I would imagine that `variant()` should imply this if the user doesn't specify `values=...`. 

OK I see... nothing means a boolean variant, we probably need to keep that.  Still, users shouldn't have to write `lambda`.  Let's at least wrap this up into a nicely-named symbol somewhere.

File: var/spack/repos/builtin/packages/npb/package.py [hunk 2]
```diff
@@ -22,9 +22,19 @@
 # License along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 ##############################################################################
+import numbers
+
 from spack import *
 
 
+def is_integral(x):
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/npb/package.py]: Would it be possible to build this more into core Spack?  Either (a) add it as a core Spack method, or (b) add true `int`, `bool`, `str`, `mutli-valued` variant types.

File: var/spack/repos/builtin/packages/npb/package.py [hunk 3]
```diff
@@ -66,20 +76,39 @@ class Npb(MakefilePackage):
...
-    variant('openmp', default=False, description='Build the OpenMP implementation')
-    variant('serial', default=False, description='Build the serial version')
-
-    # TODO: Convert these to non-exclusive multi-valued variants
-    variant('names', default=','.join(valid_names),
-            description='Benchmark names (comma separated list)')
-    variant('classes', default=','.join(valid_classes),
-            description='Benchmark classes (comma separated list)')
+    variant(
+        'implementation',
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/npb/package.py]: Let's think up a standard name for this variant that is (a) descriptive and (b) we can use everywhere.  How about `parallel-implementation` or `parallel-type` or `parallelism`?

File: var/spack/repos/builtin/packages/npb/package.py [hunk 4]
```diff
@@ -66,20 +76,39 @@ class Npb(MakefilePackage):
...
+        default='mpi',
+        values=('serial', 'mpi', 'openmp'),
+        description='Selects one among the available implementations'
+    )
+
+    variant(
+        'names',
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/npb/package.py]: Shouldn't this be called `benchmark-names`?  Also, should it be a variant at all, or an install option?

File: var/spack/repos/builtin/packages/npb/package.py [hunk 5]
```diff
@@ -66,20 +76,39 @@ class Npb(MakefilePackage):
...
+        default=','.join(valid_names),
+        values=valid_names,
+        multi=True,
+        description='Benchmark names (comma separated list)'
+    )
+
+    variant(
+        'classes',
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/npb/package.py]: Same thing... `benchmark-classes`

File: var/spack/repos/builtin/packages/matlab/package.py (hunk 1 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/matlab/package.py]: Commit comment: "	variants: fixed packages reported by @adamjstewart in #4098" I'll add it to the description.

File: var/spack/repos/builtin/packages/npb/package.py [hunk 6]
```diff
@@ -66,20 +76,39 @@ class Npb(MakefilePackage):
...
+        'implementation',
+        default='mpi',
+        values=('serial', 'mpi', 'openmp'),
+        description='Selects one among the available implementations'
+    )
+
+    variant(
+        'names',
+        default=','.join(valid_names),
+        values=valid_names,
```
Comment (hartzell) [inline var/spack/repos/builtin/packages/npb/package.py]: How does this work?  `valid_names` is a list of the possible names, but the value of the variant is going to be a string built by joining some subset of `valid_names` with commas.  Is the magic sauce [this bit here in `variant.py`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/variant.py#L221), which looks like it converts the comma-separated string back into a uniq'ed list?

File: var/spack/repos/builtin/packages/npb/package.py (hunk 2 above)
Comment (hartzell) [inline var/spack/repos/builtin/packages/npb/package.py]: Perhaps `variant.py` could provide a set of helper functions for validating variant values?

File: var/spack/repos/builtin/packages/npb/package.py (hunk 4 above)
Comment (hartzell) [inline var/spack/repos/builtin/packages/npb/package.py]: Is it neat-and-clean for a PR that's addressing a set of problems with multi-valued variants to also be renaming all of a packages variants?

File: var/spack/repos/builtin/packages/npb/package.py (hunk 3 above)
Comment (hartzell) [inline var/spack/repos/builtin/packages/npb/package.py]: ditto my variant renaming comment below.

File: var/spack/repos/builtin/packages/npb/package.py (hunk 2 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/npb/package.py]: @hartzell I was thinking about that. If I am moving the function to `variant.py` I would like to agree on naming. Is something like `is_integral` good enough (focus on the predicate) or do you prefer something like `any_integer` (more declarative)? @adamjstewart @davydden @citibeth

File: var/spack/repos/builtin/packages/npb/package.py (hunk 6 above)
Comment (alalazo) [inline var/spack/repos/builtin/packages/npb/package.py]: Yes, that part turns the values given in input into a uniquely sorted tuple of values. [This other part](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/variant.py#L115) validates against the `values=<something>` argument.

File: var/spack/repos/builtin/packages/npb/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/npb/package.py]: I like `any_integer` better than `is_integral`. But even better would be @citibeth's suggestion of `type=int`. So you could either use `values` or `type` (or neither if any value works?) to restrict the possible values.

Comment (hartzell) [inline var/spack/repos/builtin/packages/npb/package.py]: I like the idea of values being (able to be) a lamba (that seems to be the pythonic word...) that tests a value and returns True/False.  Given that, having&supporting a second knob named `type` seems redundant->more_work->confusing. 

You could think about what Perl/Moose did with [TypeConstraints](http://search.cpan.org/~ether/Moose-2.2005/lib/Moose/Util/TypeConstraints.pm), it's a fairly flexible system for adding "types", I'd imagine you might want something like that so that I could say "type=IntBetween23And42" or "type=PhoneNumber" or ....

So far we have one example of this and I fear over-engineering it.

RE a naming standard:

- What does `any_integer` do?  It returns True/False if its arg is an integer.
- What does `is_integral/is_integer` do?  It returns True/False if its arg is an integer.

I think that the second is more mnemonic/descriptive, but that's probably because the "is_" bit implies a boolean return value and perhaps that's not critical.  In fact, you're going to end up with a library of functions, all of which begin with `is_` (or `is_not_`), at which point it might loose it's value.

Any votes for `integerp`? (mostly :wink:)

Bikeshed?

File: var/spack/repos/builtin/packages/matlab/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/matlab/package.py]: In this case, the value can be any string. It would be nice if there was a way to say `values=str` or `type=str`.

File: var/spack/repos/builtin/packages/plumed/package.py [hunk 7]
```diff
@@ -51,8 +51,12 @@ class Plumed(AutotoolsPackage):
     # The ones listed here are not built by default for various reasons,
     # such as stability, lack of testing, or lack of demand.
     # FIXME: This needs to be an optional
-    variant('optional_modules', default='all',
-            description='String that is used to build optional modules')
+    variant(
+        'optional_modules',
+        default='all',
+        values=lambda x: True,
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/plumed/package.py]: There is probably actually a set list of valid strings, but someone would have to look them up.

File: var/spack/repos/builtin/packages/npb/package.py (hunk 4 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/npb/package.py]: I actually wrote this package, so all renaming arguments can be directed towards me. I still stand by my original names, although I don't really care that much. I agree with @hartzell, let's rename variants in a different PR if we decide to do that.

File: var/spack/repos/builtin/packages/npb/package.py [hunk 8]
```diff
@@ -88,23 +117,23 @@ class Npb(MakefilePackage):
...
+        if 'implementation=mpi' in self.spec:
             implementation = 'MPI'
-        elif '+openmp' in self.spec:
+        elif 'implementation=openmp' in self.spec:
             implementation = 'OMP'
-        elif '+serial' in self.spec:
+        elif 'implementation=serial' in self.spec:
             implementation = 'SER'
         else:
             raise RuntimeError('You must choose an implementation to build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/npb/package.py]: You can get rid of this `RuntimeError` now that it is physically impossible to specify an invalid implementation. That's my favorite features of multi-valued variants.

**Comment by junghans** [comment]
After rebasing #3852 on this patch set, I am getting this error:
```
==> Error: Invalid spec: 'charm@6.7.1%gcc@6.3.0 backend=mpi ~papi+shared+smp~tcp arch=linux-gentoo2-x86_64'. Package charm requires variant backend=mpi, but spec asked for backend=mpi
```

**Comment by tgamblin** [comment]
@junghans: I think you hit #4105.
